pkg/controller/utils/machine_helper_test.go

```diff
@@ -0,0 +1,280 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	awsproviderapi "sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsproviderconfig/v1beta1"
+)
+
+func TestUpdateAWSLBList(t *testing.T) {
+	clustername := "lbupdatetest"
+	sampleMachine := testutils.CreateMachineObj("master-0", clustername, "master", testutils.DefaultRegionName, testutils.DefaultAzName)
+	objs := []runtime.Object{&sampleMachine}
+	mocks := testutils.NewTestMock(t, objs)
+
+	oldLBList := []awsproviderapi.LoadBalancerReference{
+		{
+			// <clustername>-<id>-ext
+			Name: fmt.Sprintf("%s-%s-ext", clustername, "12345"),
+			Type: awsproviderapi.NetworkLoadBalancerType,
+		},
+		{
+			// <clustername>-<id>-int
+			Name: fmt.Sprintf("%s-%s-int", clustername, "12345"),
+			Type: awsproviderapi.NetworkLoadBalancerType,
+		},
+	}
+	newLBList := []awsproviderapi.LoadBalancerReference{
+		{
+			// Just something random!
+			Name: fmt.Sprintf("%s-%s-test", clustername, "12345"),
+			Type: awsproviderapi.NetworkLoadBalancerType,
+		},
+	}
+	// quickly make sure the test is going to measure an actual change
+	if len(oldLBList) != 2 || len(newLBList) != 1 {
+		t.Fatalf("Initial test conditions are unexpected. Old LB list should be 2 (got %d), New LB list should be 1 (got %d)", len(oldLBList), len(newLBList))
+	}
+	// decode spec
+	codec, err := awsproviderapi.NewCodec()
+	if err != nil {
+		t.Fatalf("Can't create decoder codec for AWS Provider API %v", err)
+	}
+	awsconfig := &awsproviderapi.AWSMachineProviderConfig{}
+	err = codec.DecodeProviderSpec(&sampleMachine.Spec.ProviderSpec, awsconfig)
+	if err != nil {
+		t.Fatalf("Can't decode sample Machine ProviderSpec %v", err)
+	}
+
+	err = updateAWSLBList(mocks.FakeKubeClient, oldLBList, newLBList, sampleMachine, awsconfig)
+	if err != nil {
+		t.Fatalf("Couldn't update AWS LoadBalancer List: %v", err)
+	}
+
+	machineInfo := types.NamespacedName{
+		Name:      sampleMachine.GetName(),
+		Namespace: sampleMachine.GetNamespace(),
+	}
+	// reload the object to make sure we're not just working with the "in-memory"
+	// representation, that being, the un-saved one.
+	err = mocks.FakeKubeClient.Get(context.TODO(), machineInfo, &sampleMachine)
+	if err != nil {
+		t.Fatalf("Couldn't reload the test machine: %v", err)
+	}
+
+	err = codec.DecodeProviderSpec(&sampleMachine.Spec.ProviderSpec, awsconfig)
+	if err != nil {
+		t.Fatalf("Can't decode sample Machine ProviderSpec %v", err)
+	}
+
+	if len(awsconfig.LoadBalancers) != 1 {
+		t.Fatalf("Expected to have only 1 LoadBalancer, but got %d", len(awsconfig.LoadBalancers))
+	}
+	if awsconfig.LoadBalancers[0].Name != newLBList[0].Name {
+		t.Fatalf("Expected LB name to be %s, but got %s", newLBList[0].Name, awsconfig.LoadBalancers[0].Name)
+	}
+
+	if awsconfig.LoadBalancers[0].Type != newLBList[0].Type {
+		t.Fatalf("Expected LB type to be %s, but got %s", string(newLBList[0].Type), string(awsconfig.LoadBalancers[0].Type))
+	}
+}
```

Comment:
> we may also want a test here to ensure that no change is made if the two LB lists are the same

Wouldn't this test still handle it? If `newLBList` is treated as the authoritative set of load balancers for the machines in question we would expect the code to always ensure that the machine load balancers are equal to that set. If it happens that `newLBList` and oldLBList` are the same, isn't the `updateAWSLBList` code effectively idempotent because at the end of the worst case, the machine load balancers will be removed and re-added with the same values?